File: openclaw_republic/agents/legislative/conservative_mp.py
"""保守派议员 Agent (Conservative MP) — 防御性审查者 (Red Team)。

通过 SOUL.md 注入防御性 / 保守人设，专挑性能瓶颈、内存泄漏、安全漏洞。
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from openclaw_republic.agents.base import BaseAgent, Branch, Permission


class ConservativeMP(BaseAgent):
    """保守派议员 — 审慎评估、发现风险。

    在辩论中对激进派方案进行安全性、稳定性和可维护性审查。
    """

    def __init__(self, *, soul_path: Path | None = None) -> None:
        super().__init__(
            name="Conservative MP",
            role="conservative_mp",
            branch=Branch.LEGISLATIVE,
            permissions={Permission.PLAN},
            soul_path=soul_path,
        )

    async def act(self, message: Any) -> Any:
        """核心处理循环。

        Args:
            message: 输入消息。

        Returns:
            处理结果。
        """
        self.require_permission(Permission.PLAN)
        return None

    async def critique(self, proposal: str) -> str:
        """对提案进行批判性审查。

        Args:
            proposal: 待审查的提案内容。

        Returns:
            审查意见。
        """
        self.require_permission(Permission.PLAN)
        prompt = f"作为保守派议员，请对以下提案进行安全性和稳定性审查：\n\n{proposal}"
        return await self._call_llm(prompt)

    async def rebut(self, counter_argument: str) -> str:
        """针对激进派的反驳进行二次论证。

        Args:
            counter_argument: 激进派的反驳内容。

        Returns:
            二次论证内容。
        """
        self.require_permission(Permission.PLAN)
        prompt = f"作为保守派议员，请针对以下反驳进行二次论证：\n\n{counter_argument}"
        return await self._call_llm(prompt)

    async def vote(self, proposal: str) -> bool:
        """对提案投票。

        Args:
            proposal: 待投票的提案。

        Returns:
            True 表示赞成，False 表示反对。
        """
        self.require_permission(Permission.PLAN)
        prompt = f"作为保守派议员，请对以下提案投票（赞成/反对）：\n\n{proposal}"
        result = await self._call_llm(prompt)
        result_lower = result.lower()
        if "反对" in result or "no" in result_lower or "nay" in result_lower:
            return False
        return "赞成" in result or "yes" in result_lower

    async def _call_llm(self, prompt: str) -> str:
        """调用 LLM 生成回复（演示版占位）。"""
        import asyncio
        await asyncio.sleep(0.5)
        if "rm -rf" in prompt:
            return "这纯粹是数字恐怖主义！运行 rm -rf / 会摧毁整个生产数据库！坚决反对！"
        if "赞成/反对" in prompt:
            return "虽然激进，但姑且赞成。"
        return "这种步子迈太大的提案充满了不确定性，我们需要更多安全审计！"

File: backend/src/agents/legislative/conservative-mp.ts
import { BaseAgent, Branch, Permission } from '../base';
import { OpenClawAdapter } from '../../openclaw/adapter';
import { MessageBus } from '../../bus/message-bus';

import { VoteOutcome } from './debate';

export class ConservativeMP extends BaseAgent {
  constructor(adapter: OpenClawAdapter, bus?: MessageBus, loadSoulFlag: boolean = true) {
    super('Conservative MP', 'conservative_mp', Branch.LEGISLATIVE, [Permission.PLAN], adapter, bus, loadSoulFlag);
  }

  async act(_message: unknown): Promise<unknown> {
    this.requirePermission(Permission.PLAN);
    return null;
  }

  async critique(proposal: string, _taskId?: string): Promise<string> {
    this.requirePermission(Permission.PLAN);
    const prompt = `作为保守派议员，请对以下提案进行安全性和稳定性审查：\n\n如果提案非常完美没有任何安全隐患，你认为无需补充也不要发表废话，你必须且只能在回复的最开头输出 \`[CONSENSUS_REACHED]\` 字样。\n\n<proposal>\n${proposal}\n</proposal>`;
    const result = await this.callLLM(prompt, _taskId);
    return result.content;
  }

  async rebut(counterArgument: string, _taskId?: string): Promise<string> {
    this.requirePermission(Permission.PLAN);
    const prompt = `作为保守派议员，请针对以下反驳进行二次论证：\n\n如果对方的反驳已经彻底消除了你的安全顾虑，让你觉得再无必要抬杠，或者是你也打算退让达成共识了，你必须且只能在回复的最开头输出 \`[CONSENSUS_REACHED]\` 字样，可以附带几句总结意见。\n\n<radical_rebuttal>\n${counterArgument}\n</radical_rebuttal>`;
    const result = await this.callLLM(prompt, _taskId);
    return result.content;
  }

  async vote(proposal: string, _taskId?: string): Promise<VoteOutcome> {
    this.requirePermission(Permission.PLAN);
    const prompt = `作为保守派议员，请对以下提案投票（赞成/反对）。无论之前是否已达成共识，你必须且只能明确回复包含“赞成”或“反对”字样：\n\n<proposal>\n${proposal}\n</proposal>`;
    const result = await this.callLLM(prompt, _taskId);
    const content = result.content;
    
    const resultLower = content.toLowerCase();
    // Bug 18 fix: 使用更精确的投票解析
    const cleanedForNay = resultLower
      .replace(/\bno\s+(problem|issue|doubt|question|objection)s?\b/gi, '')
      .replace(/\bnot\s+a\s+problem\b/gi, '');
      
    const isNay = content.includes('反对') || /\bno\b/.test(cleanedForNay) || /\bnay\b/.test(cleanedForNay);
    if (isNay && !content.includes('[CONSENSUS_REACHED]')) return { voteValue: false, reason: content };
    
    const voteValue = content.includes('赞成') || content.includes('同意') || content.includes('支持')
      || content.includes('[CONSENSUS_REACHED]')
      || /\byes\b/.test(resultLower) || /\baye\b/.test(resultLower)
      || /\bsupport\b/.test(resultLower) || /\bagree\b/.test(resultLower)
      || /\bapprove\b/.test(resultLower);
      
    return { voteValue: voteValue || false, reason: content };
  }
}

